src/Motor.Extensions.Hosting.Kafka/KafkaMessageConsumer.cs

```diff
@@ -162,47 +166,43 @@ private void WriteStatistics(string json)
         }
     }
 
-    private void SingleMessageHandling(CancellationToken token, ConsumeResult<string?, byte[]> msg)
+    private async Task<bool> SingleMessageHandlingAsync(CancellationToken token, ConsumeResult<string?, byte[]> msg)
     {
-        _logger.LogDebug(
-            $"Received message from topic '{msg.Topic}:{msg.Partition}' with offset: '{msg.Offset}[{msg.TopicPartitionOffset}]'");
+        _logger.LogDebug("Received message from topic '{Topic}:{Partition}' with offset: '{Offset}[{TopicPartitionOffset}]'",
+            msg.Topic, msg.Partition, msg.Offset, msg.TopicPartitionOffset);
         var cloudEvent = KafkaMessageToCloudEvent(msg.Message);
 
-        var taskAwaiter = ConsumeCallbackAsync?.Invoke(cloudEvent, token).GetAwaiter();
-        taskAwaiter?.OnCompleted(() =>
+        return await await ConsumeCallbackAsync?.Invoke(cloudEvent, token)
+            .ContinueWith(async status => HandleMessageStatus(msg, await status), token)!;
+    }
+
+    private bool HandleMessageStatus(ConsumeResult<string?, byte[]> msg, ProcessedMessageStatus? status)
+    {
+        try
         {
-            var processedMessageStatus = taskAwaiter?.GetResult();
-            _messageSemaphore.Release();
-            switch (processedMessageStatus)
+            switch (status)
             {
                 case ProcessedMessageStatus.Success:
-                    break;
-                case ProcessedMessageStatus.TemporaryFailure:
-                    break;
                 case ProcessedMessageStatus.InvalidInput:
-                    break;
+                    if (msg.Offset.Value % _options.CommitPeriod == 0)
+                    {
+                        _consumer?.Commit(msg);
+                    }
+                    _messageSemaphore.Release();
+                    return true;
+                case ProcessedMessageStatus.TemporaryFailure:
                 case ProcessedMessageStatus.CriticalFailure:
-                    break;
                 case ProcessedMessageStatus.Failure:
-                    break;
+                    return false;
```

Comment:
I'd argue that it might not be ideal to retry whenever we get a `CriticalFailure` or `Failure`. In the first case, the RabbitConsumer stops the application, I think this would also be good here. In the latter case, Rabbit just rejects, but I think it would be best to also stop, as I find it counterintuitive that it retries if it fails. Only in case of `TemporaryFailure` we should retry.